ee/agent/reset.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	serial, ok := resp[0]["hardware_serial"]
+	if !ok {
+		return "", "", errors.New("hardware_serial missing from results")
+	}
+
+	hardwareUUID, ok := resp[0]["hardware_uuid"]
+	if !ok {
+		return "", "", errors.New("hardware_uuid missing from results")
+	}
+
+	return serial, hardwareUUID, nil
+}
+
+// valueChanged checks the knapsack for the given data and compares it to
+// currentValue. A value is considered changed only if the stored value was
+// previously set.
+func valueChanged(k types.Knapsack, currentValue string, dataKey []byte) bool {
+	storedValue, err := k.HostDataStore().Get(dataKey)
+	if err != nil {
+		k.Slogger().Log(context.TODO(), slog.LevelError, "could not get stored value", "err", err, "key", string(dataKey))
+		return false // assume no change
+	}
+
+	if len(storedValue) == 0 {
+		k.Slogger().Log(context.TODO(), slog.LevelDebug, "value not previously stored, storing now", "key", string(dataKey))
+		if err := k.HostDataStore().Set(dataKey, []byte(currentValue)); err != nil {
+			k.Slogger().Log(context.TODO(), slog.LevelError, "could not store value", "err", err, "key", string(dataKey))
+		}
+		return false
+	}
+
+	if storedValue != nil && currentValue != string(storedValue) {
+		k.Slogger().Log(context.TODO(), slog.LevelInfo, "hardware-identifying value has changed", "key", string(dataKey))
+		return true
+	}
+
+	return false
+}
+
+// currentMunemo retrieves the enrollment secret from either the knapsack or the filesystem,
+// depending on launcher configuration, and then parses the tenant munemo from it.
+func currentMunemo(k types.Knapsack) (string, error) {
+	var enrollSecret string
+	if k.EnrollSecret() != "" {
+		enrollSecret = k.EnrollSecret()
+	} else if k.EnrollSecretPath() != "" {
+		content, err := os.ReadFile(k.EnrollSecretPath())
+		if err != nil {
+			return "", fmt.Errorf("could not read secret at enroll_secret_path %s: %w", k.EnrollSecretPath(), err)
+
+		}
+		enrollSecret = string(bytes.TrimSpace(content))
+	} else {
+		return "", errors.New("enroll secret and secret path both unset")
+	}
+
+	// We cannot verify since we don't have the key, so we use ParseUnverified
+	token, _, err := new(jwt.Parser).ParseUnverified(enrollSecret, jwt.MapClaims{})
+	if err != nil {
+		return "", fmt.Errorf("parsing enroll secret jwt: %w", err)
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return "", errors.New("enroll secret has no claims")
+	}
+
+	munemo, ok := claims["organization"]
+	if !ok {
+		return "", errors.New("enroll secret claims missing organization claim")
+	}
+
+	munemoStr, ok := munemo.(string)
+	if !ok {
+		return "", errors.New("munemo is unsupported type")
+	}
+
+	return munemoStr, nil
+}
+
+// takeDatabaseBackup retrieves the data we want to preserve from various db stores
+// as a snapshot of this db, appends it to previous snapshots if they exist, and
+// returns the collection of backup data.
+func takeDatabaseBackup(k types.Knapsack) ([]byte, error) {
```

Comment:
I agree that `take backup` is misleading. I don't want to overload `tombstone` with the k2 meaning (esp since we have a tombstone_id inside the returned data), and a database record sounds like a single item in a database rather than a collection of items to me. `snapshot`, which I use in the docblock and I think in the tests too, feels a little less incorrect but still misleading.

Maybe `prepareDatabaseResetRecord`? Now that we have both a reset reason and a reset timestamp, maybe it makes more sense to rename the `oldHostData` struct to `dbResetRecord`.

Going to go with that for now, but let me know if you have other ideas.

Reply:
I'm thinking about the method name, and it might make sense to change anything else. Maybe `oldDatabaseRecord` ? `dbWipeRecord` or `dbResetRecord` are good too. 

Maybe I like `dbResetRecord` best. Or maybe it's all good enough